sirepo/template/sdds_util.py

```diff
@@ -41,6 +42,90 @@
 _sdds_lock = threading.RLock()
 
 
+class SDDSUtil:
+    def __init__(self, filename):
+        self.filename = filename
+
+    def heatmap(self, plot_attrs):
+        sdds_label_and_units = {}
+        x_col = extract_sdds_column(self.filename, plot_attrs.x, 0)
+        if x_col.err:
+            return x_col.err
+        x_values = x_col["values"]
+        sdds_label_and_units["x"] = (x_col.column_def[0], x_col.column_def[1])
+        y_col = extract_sdds_column(self.filename, plot_attrs.y, 0)
+        if y_col.err:
+            return y_col.err
+        y_values = y_col["values"]
+        sdds_label_and_units["y"] = (y_col.column_def[0], y_col.column_def[1])
+        model = plot_attrs.frame_args.sim_in.models[plot_attrs.frame_args.frameReport]
+        model.update(plot_attrs.frame_args)
+
+        plot_fields = PKDict()
+        for f in ("x", "y"):
+            plot_attrs.format_plot(
+                plot_fields,
+                getattr(plot_attrs, f),
+                sdds_label_and_units[f][0],
+                sdds_label_and_units[f][1],
+            )
+
+        return template_common.heatmap(
+            values=[x_values, y_values],
+            model=model,
+            plot_fields=PKDict(
+                x_label=plot_fields["x_label"],
+                y_label=plot_fields["y_label"],
+                title=plot_attrs.title,
+            ),
+        )
+
+    def lineplot(self, plot_attrs):
+        x = None
+        plots = []
+        for f in plot_attrs.x + plot_attrs.y:
+            if f not in plot_attrs.frame_args or plot_attrs.frame_args[f] == "none":
+                continue
+            col_name = plot_attrs.format_col_name(plot_attrs.frame_args[f])
+            col = extract_sdds_column(self.filename, col_name, 0)
+            if col.err:
+                return col.err
+            plot = PKDict(
+                points=col["values"],
+                label=plot_attrs.frame_args[f],
+                col_name=col_name,
+            )
+            if f == "x":
+                plot.x = True
```

Comment:
I don't like this, but I was trying to find a way to handle the different case for formatting x in _sdds_report

Reply:
This is good progress. Here are a few items which should clean it up a bit:

The API to line plot is called like this:
```python
    return sdds_util.SDDSUtil(str(frame_args.run_dir.join(_OPAL_SDDS_FILE))).lineplot(
        PKDict(
            x=("x",),
            y=("y1", "y2", "y3"),
            ...
        )
    )
```
That may confuse people into thinking there could multiple "x" values allowed which isn't the case. Instead, the API should probably be changed to something like this, which is very explicit that there is one x values allowed and multiple y values.
```python
    return sdds_util.SDDSUtil(str(frame_args.run_dir.join(_OPAL_SDDS_FILE))).lineplot(
        PKDict(
            x_field="x",
            y_fields=("y1", "y2", "y3"),
            ...
        )
    )
```

I've done some testing with jspec and the bug related to sorting vtrans seems to be fixed in the physics code, and I can't find a case where the V_trans values are out of order. So we can remove ```def _resort_vtrans(...)``` from template.jspec and remove the ```v_trans_hook``` parameter from SDDSUtil.

The code 
```python
    if f == "x":
        plot.x = True
```
can be removed from SDDSUtil if the line is changed in template.jspec from:
```python
    def _format_plot(plot, sdds_units):
        if "x" in plot and plot.x:
            ....
```
to
```python
    def _format_plot(plot, sdds_units):
        if x_field == plot.col_name:
            ....
```

also, in SDDSUtil, the line:
```python
            if f == "x":
                x = plot
```
could be changed to this, avoiding the static literal:
```python
            # after plot_attrs.x is renamed to plot.attrs.x_field
            if f == plot_attrs.x_field:
                x = plot
```

Reply:
bc2d5a6844bd5910a78960ca4e509e4f80fa3deb